jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/util/WebSocketDemander.java

```diff
@@ -97,21 +141,43 @@ public void setNextDemand(DemandChain nextDemand)
      */
     public void onFrame(Frame frame, Callback callback)
     {
-        if (_frame != null || _callback != null)
-            throw new IllegalStateException("Not expecting onFrame");
+        Throwable failure = null;
+        try (AutoLock l = _lock.lock())
+        {
+            if (_failure != null)
+            {
+                failure = _failure;
+            }
+            else if (_inputDemand)
+            {
+                // Fulfill the incoming demand to receive a new frame.
+                _inputDemand = false;
+                _frame = frame;
+                _callback = new CountingCallback(callback, 1);
+            }
+            else
+            {
+                failure = new IllegalStateException("No demand for onFrame");
+            }
+        }
 
-        _frame = frame;
-        _callback = new CountingCallback(callback, 1);
-        succeeded();
+        if (failure == null)
+        {
+            succeeded();
+        }
+        else
+        {
+            callback.failed(failure);
+            failFlusher(failure);
+        }
```

Comment:
This feels wrong, because if no failure, the whole `IteratingCallback` is succeeded, but not the `callback` parameter, and if there is a failure, the `callback` parameter is failed, but not the `IteratingCallback` (which is instead aborted).

Reply:
We can't succeed the callback parameter because we are not done with the frames payload, but if we have a non-null `failure` we need to fail it because it wasn't added to the state and so won't be failed later on.

I don't think we want to just fail the iterating callback. It is really a terminal failure so we want to ensure that the Demander is aborted. Looking at ICB.failed there are cases where it just turns into a NOOP, for example if it has already been completed.

Reply:
Please add a comment to the call to `succeeded()`, something like:

```
// Succeed this IteratingCallback, since we are not done with the frame's payload.